test/e2e/operator_test.go

```diff
@@ -228,3 +233,402 @@ func setImage(deployment *appsv1.Deployment, image string) {
+		t.Fatalf("failed to create configmap %s/%s: %v", upstreamCfgMap.Namespace, upstreamCfgMap.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), upstreamCfgMap); err != nil {
+			t.Fatalf("failed to delete configmap %s/%s: %v", upstreamCfgMap.Namespace, upstreamCfgMap.Name, err)
+		}
+	}()
+
+	// Get the CoreDNS image used by the test upstream resolver.
+	co := &configv1.ClusterOperator{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Name: operatorcontroller.DNSOperatorName}, co); err != nil {
+		t.Fatalf("failed to get cluster operator dns: %v", err)
+	}
+	var upstreamImage string
+	for _, ver := range co.Status.Versions {
+		if ver.Name == "coredns" {
+			if len(ver.Version) == 0 {
+				t.Fatalf("dns operator does not contain a coredns version: %v", err)
+			}
+			upstreamImage = ver.Version
+			break
+		}
+	}
+
+	// Create the upstream resolver Pod.
+	upstreamResolver := upstreamPod(upstreamPodName, upstreamPodNS, upstreamImage, upstreamPodName)
+	if err := cl.Create(context.TODO(), upstreamResolver); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), upstreamResolver); err != nil {
+			t.Fatalf("failed to delete pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+		}
+	}()
+
+	// Wait for the upstream resolver Pod to be ready.
+	err = wait.PollImmediate(1*time.Second, 60*time.Second, func() (bool, error) {
+		if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: upstreamResolver.Namespace, Name: upstreamResolver.Name}, upstreamResolver); err != nil {
+			return false, nil
+		}
+		for _, cond := range upstreamResolver.Status.Conditions {
+			if cond.Type == corev1.ContainersReady &&
+				cond.Status == corev1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to observe ContainersReady condition for pod %s/%s: %v", upstreamResolver.Namespace, upstreamResolver.Name, err)
+	}
+
+	// Create the upstream resolver Service and get the ClusterIP.
+	upstreamSvc := upstreamService(upstreamPodName, upstreamPodNS)
+	if err := cl.Create(context.TODO(), upstreamSvc); err != nil {
+		t.Fatalf("failed to create service %s/%s: %v", upstreamSvc.Namespace, upstreamSvc.Name, err)
+	}
+	defer func() {
+		if err := cl.Delete(context.TODO(), upstreamSvc); err != nil {
+			t.Fatalf("failed to delete service %s/%s: %v", upstreamSvc.Namespace, upstreamSvc.Name, err)
+		}
+	}()
+	if err := cl.Get(context.TODO(), types.NamespacedName{upstreamSvc.Namespace, upstreamSvc.Name}, upstreamSvc); err != nil {
+		t.Fatalf("failed to get service %s/%s: %v", upstreamSvc.Namespace, upstreamSvc.Name, err)
+	}
+	upstreamIP := upstreamSvc.Spec.ClusterIP
+
+	// Update cluster DNS forwarding with the upstream resolver's Service IP address.
+	defaultDNS := &operatorv1.DNS{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
I don't think you need to initialize this struct

Reply:
ack

Reply:
ack